subsys/dfu/fmfu_fdev/src/fmfu_fdev.c

```diff
@@ -0,0 +1,260 @@
+			size_t *segments_len, Segments_t *segments)
+{
+	if (!cbor_decode_Segments(segments_buf, buf_len, segments, NULL)) {
+		return -EINVAL;
+	}
+
+	*segments_len = 0;
+	for (int i = 0; i < segments->_Segments__Segment_count; i++) {
+		*segments_len += segments->_Segments__Segment[i]._Segment_len;
+	}
+
+	return 0;
+}
+
+static int load_segment(const struct device *fdev, size_t seg_size,
+			uint32_t seg_target_addr, uint32_t seg_offset)
+{
+	int err;
+	uint32_t read_addr = seg_offset;
+	size_t bytes_left = seg_size;
+
+	while (bytes_left) {
+		uint32_t read_len = MIN(NRF_FMFU_MODEM_BUFFER_SIZE, bytes_left);
+
+		err = flash_read(fdev, read_addr, buf, read_len);
+		if (err != 0) {
+			LOG_ERR("flash_read failed2: %d", err);
+			return err;
+		}
+
+		err = nrf_fmfu_memory_chunk_write(seg_target_addr, read_len,
+						  buf);
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_memory_chunk_write failed: %d", err);
+			return err;
+		}
+
+		LOG_DBG("wrote chunk: offset 0x%x target addr 0x%x size 0x%x",
+			read_addr, seg_target_addr, read_len);
+
+		seg_target_addr += read_len;
+		bytes_left -= read_len;
+		read_addr += read_len;
+	}
+
+	return 0;
+}
+
+static int load_segments(Segments_t *seg, size_t blob_offset)
+{
+	int err;
+	size_t prev_segments_len = 0;
+
+	LOG_INF("Writing %d segments", seg->_Segments__Segment_count);
+
+	for (int i = 0; i < seg->_Segments__Segment_count; i++) {
+		size_t seg_size = seg->_Segments__Segment[i]._Segment_len;
+		uint32_t seg_addr =
+			seg->_Segments__Segment[i]._Segment_target_addr;
+		uint32_t read_addr = blob_offset + prev_segments_len;
+
+		err = nrf_fmfu_transfer_start();
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_transfer_start failed1: %d", err);
+			return err;
+		}
+
+		err = load_segment(fdev, seg_size, seg_addr, read_addr);
+		if (err != 0) {
+			LOG_ERR("load_segment failed: %d", err);
+			return err;
+		}
+
+		err = nrf_fmfu_transfer_end();
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_transfer_end failed: %d", err);
+			return err;
+		}
+
+		prev_segments_len += seg_size;
+
+		LOG_INF("Segment %d written. Target addr: 0x%x, size: 0%x",
+			i, seg_addr, seg_size);
+	}
+
+	LOG_INF("FMFU finished");
+
+	return 0;
+}
+
+int fmfu_fdev_init(uint8_t *buf_in, size_t buf_len_in,
+		 const struct device *fdev_in)
+{
+	if (buf_in == NULL || fdev_in == NULL) {
+		return -EINVAL;
+	}
+
+	buf = buf_in;
+	buf_len = buf_len_in;
+	fdev = fdev_in;
+
+	return 0;
+}
+
+int fmfu_fdev_load(size_t offset)
+{
+	const cbor_string_type_t *segments_string;
+	COSE_Sign1_Manifest_t wrapper;
+	bool hash_len_valid = false;
+	uint8_t expected_hash[32];
+	bool hash_valid = false;
+	bool sig_valid = false;
+	Segments_t segments;
+	size_t blob_offset;
+	uint8_t hash[32];
+	size_t blob_len;
+	int err;
+
+	if (buf == NULL || fdev == NULL) {
+		LOG_ERR("Not initialized, please call fmfu_fdev_init()");
+		return -ENOMEM;
+	}
+
+	/* Put modem in DFU/RPC state */
+	err = nrf_fmfu_init(NULL, buf_len, buf);
```

Comment:
Are you referring the use of NULL?

This is done to skip fetching the digest. You only need the digest in a wired scenario when you can fail and start over again.

Reply:
`buf` is first passed to nrf_fmfu_init here, then to flash_read and cbor_decode_Wrapper below. Since cbor_decode_Wrapper doesn't copy strings, `buf` is in use as long as `wrapper` is in use. Maybe nrf_fmfu_init can be called in load_segments instead?

Reply:
Moving _buf to only be used as parameter for _load so this should be resolved now.